test/references/stream_reference_test.cpp

```diff
@@ -1,66 +1,100 @@
-/** @file test/generators/randoms_test.cpp
- *  @author Kyle Krol */
+/** @file test/references/stream_reference_test.cpp
+ *  @author Kyle Krol
+ */
 
 #include <lin/core.hpp>
-#include <lin/references/stream_reference.hpp>
+#include <lin/references.hpp>
 
 #include <gtest/gtest.h>
 
 #include <type_traits>
 
-TEST(StreamReference, Reference) {
-  lin::Matrixf<3, 3> A({
+TEST(StreamReference, MatrixStreamReference) {
+  lin::Matrix3x3f const A = {
     0.0f, 1.0f, 2.0f,
     3.0f, 4.0f, 5.0f,
     6.0f, 7.0f, 8.0f
-  });
-  lin::Matrixf<2, 2> B({
+  };
+  lin::Matrix2x2f const B = {
     3.0f, 4.0f,
     6.0f, 7.0f
-  });
-  auto const C = lin::ref<2, 2>(A, 1, 0);
+  };
+
+  auto const C = lin::ref<lin::Matrix2x2f>(A, 1, 0);
+  static_assert(lin::internal::have_same_traits<std::remove_const_t<decltype(C)>, lin::Matrix2x2f>::value, "");
   ASSERT_FLOAT_EQ(0.0f, lin::fro(B - C));
   ASSERT_EQ(2, C.rows());
   ASSERT_EQ(2, C.cols());
-  ASSERT_EQ(4, C.size());
 
-  lin::Matrixf<0, 0, 5, 5> D(2, 3, {
+  lin::Matrixf<0, 0, 5, 5> const D(2, 3, {
     0.0f, 1.0f, 2.0f,
     3.0f, 4.0f, 5.0f
   });
-  lin::Matrixf<0, 0, 5, 5> E(2, 2, {
+  lin::Matrixf<0, 0, 5, 5> const E(2, 2, {
     1.0f, 2.0f,
     4.0f, 5.0f
   });
-  auto const F = lin::ref<0, 0, 5, 5>(D, 0, 1, 2, 2);
+
+  auto const F = lin::ref<lin::Matrixf<0, 0, 5, 5>>(D, 0, 1, 2, 2);
+  static_assert(lin::internal::have_same_traits<std::remove_const_t<decltype(F)>, lin::Matrixf<0, 0, 5, 5>>::value, "");
```

Comment:
why does the above not use syntax where `lin::Matrix3x3f` for example has the dim info embedded in the type?

Reply:
The use of variably sized types isn't as common throughout our code base. However, the type `lin::Matrixf<0, 0, 5, 5>` is a matrix that can have up to five rows and five columns. In this case it's set to two rows and two columns.

`lin::Matrix3x3f` and other typedefs are convenient aliases for `lin::Matrix<3, 3, 3, 3>` and others.